src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class LambdaRewriter
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        internal sealed partial class Analysis : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            private sealed class Scope
+            {
+                public Scope Parent { get; }
+
+                public ArrayBuilder<Scope> NestedScopes { get; } = ArrayBuilder<Scope>.GetInstance();
+
+                public ArrayBuilder<Closure> Closures { get; } = ArrayBuilder<Closure>.GetInstance();
+
+                public ArrayBuilder<Symbol> DeclaredCapturedVariables { get; } = ArrayBuilder<Symbol>.GetInstance();
+
+                public BoundNode BoundNode { get; }
+
+                public ArrayBuilder<LambdaFrame> Environments { get; } = ArrayBuilder<LambdaFrame>.GetInstance();
+
+                public Closure ContainingClosure { get; }
+
+                public Scope(Scope parent, BoundNode boundNode, Closure containingClosure)
+                {
+                    Parent = parent;
+                    BoundNode = boundNode;
+                    ContainingClosure = containingClosure;
+                }
+
+                public void Free()
+                {
+                    foreach (var scope in NestedScopes)
+                    {
+                        scope.Free();
+                    }
+                    NestedScopes.Free();
+
+                    foreach (var closure in Closures)
+                    {
+                        closure.Free();
+                    }
+                    Closures.Free();
+                    DeclaredCapturedVariables.Free();
+                    Environments.Free();
+                }
+            }
+
+            private sealed class Closure
+            {
+                public MethodSymbol Symbol { get; }
+
+                public PooledHashSet<Symbol> CapturedVariables { get; } = PooledHashSet<Symbol>.GetInstance();
+
+                public ArrayBuilder<LambdaFrame> CapturedEnvironments { get; } = ArrayBuilder<LambdaFrame>.GetInstance();
+
+                public NamedTypeSymbol ContainingType { get; set; }
+
+                public Closure(MethodSymbol symbol)
+                {
+                    Symbol = symbol;
+                }
+
+                public void Free()
+                {
+                    CapturedVariables.Free();
+                    CapturedEnvironments.Free();
+                }
+            }
+
+            /// <summary>
+            /// Optimizes local functions such that if a local function only references other local functions without closures, it itself doesn't need a closure.
```

Comment:
Does `closure` here specifically mean a lambda closure or does it mean the more general "captured variables"?